solutions/usaco-895.mdx

```diff
@@ -2,16 +2,96 @@
 id: usaco-895
 source: USACO Silver 2019 January
 title: Icy Perimeter
-author: Tanish Tyagi
+author: Tanish Tyagi, Brad Ma
 ---
 
-[Official Analysis](http://www.usaco.org/current/data/sol_perimeter_silver_jan19.html)
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_perimeter_silver_jan19.html)
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
+	
+<JavaSection>
+	
+```java
+import java.io.*;
+import java.util.*;
 
-<CPPSection>
+public class IcyPerimeter {
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("perimeter");
+		n = io.nextInt();
+		numberedMatrix = new int[n + 2][n + 2];
+
+		for (int x = 1; x <= n; x++) { // read matrix
+			String line = io.next();
+			for (int y = 1; y <= n; y++) {
+				String character = Character.toString(line.charAt(y - 1));
+				if (character.equals("#")) {
+					numberedMatrix[x][y] = -1; // mark # as -1
+				}
+			}
+		}
+
+		// mark each connected component with a different number
+		int markNumber = 1;
+		for (int x = 1; x <= n; x++) {
+			for (int y = 1; y <= n; y++) {
+				if(numberedMatrix[x][y] == -1) { // not visited
+					floodfill(x, y, markNumber);
+					markNumber++;
+				}
+			}
+		}
+
+		int[] rowMovement = {0, 1, 0, -1}; // dx dy arrays
```

Comment:
make these constants

Reply:
i dont think thats necessary its better to see the movements in the array than variables

Reply:
no i mean like
```java
public final int[] R_CHANGE = {0, 1, 0, -1};
```
or something